### ISSUE TITLE
`C416` triggers leads to `C409` bad preview behavior - false positive

### ISSUE THREAD

**Issue by Skylion007** [astral-sh/ruff#16500]
### Summary

Ruff triggers bad preview behavior of C409. This behavior we chose not to enable because of issue: https://github.com/astral-sh/ruff/issues/12912 . This is definitely a false positive in this case and should not be reported here as it does lead to a significant performance regression. See this issue for discussion: https://github.com/pytorch/pytorch/pull/148412

Ruff playground for the rule here

https://play.ruff.rs/898a116a-2d60-4afc-b081-a83f5ea72bf8

cc @jansel for finding and flagging this issue

At worst, it should simplify to tuple(range(10)) instead of a useless x for x in range generator

**Comment by ntBre** [comment]
Ah, without preview and accepting autofixes in the playground the code goes through these iterations:

```python
tuple([x for x in range(10)])  # initial case, C416
```

```python
tuple(list(range(10)))  # after one fix, C414
```

```python
tuple(range(10))  # after two fixes, no error
```

but with `preview = true`, C409 is also flagged and accepting that autofix leads to the problematic generator version.

Based on the name [`unnecessary-literal-within-tuple-call`](https://docs.astral.sh/ruff/rules/unnecessary-literal-within-tuple-call/#unnecessary-literal-within-tuple-call-c409), I think it might make sense if C409 were restricted to actual literals, not to comprehensions, which are covered by C416.

**Comment by Skylion007** [comment]
@ntbre this works for the case where the generator is trivial and come from ranges, but it's still actually slower when a list comprehension is transformed to a generator that cannot be further simplified. See the PyTorch issue and relevant playground for more complicated examples that won't be simplified.

The range thing I provided was just an example, but it still breaks down if the range if filtered at all or chained etc.

**Comment by MichaReiser** [comment]
I'm sorry, but I'm struggling to follow this issue. 

Checking the following file with flake8 (`uvx --with flake8-comprehensions flake8 comprehension.py`)

```py
tuple([x for x in range(10)])
```

gives me


```
comprehension.py:2:7: C416 Unnecessary list comprehension - rewrite using list().
```

Which is the same as what you get with Ruff. 

I do understand that C416 and C409 overlap and that the fix suggested by C409 is suboptimal, but that's covered in its own issue.

So is the question to change the C416 rule or is it only about that C416 and C409 don't play nicely together right now but that would be fixed by reverting the preview behavior changes in C409 (or making it its own rule)

**Comment by alessio-locatelli** [comment]
By the way, `tuple()` doesn't accept an async generator — only synchronous iterables. `C409` gives false-positives. Perhaps this deserves a dedicated issue because I cannot find related comments about this.

**Comment by ntBre** [comment]
@alessio-locatelli That does sound like a separate bug, but I think I can reproduce, if I'm following correctly:

```console
❯ ruff check --select C409 --preview --unsafe-fixes --diff - <<EOF
async def gen(): yield 1
tuple([x async for x in gen()])
EOF
@@ -1,2 +1,2 @@
 async def gen(): yield 1
-tuple([x async for x in gen()])
+tuple(x async for x in gen())
```

```pycon
>>> async def gen(): yield 1
...
>>> async def foo(): return tuple([x async for x in gen()])
...
>>> import asyncio
>>> asyncio.run(foo())
(1,)
>>> async def bar(): return tuple(x async for x in gen())
...
>>> asyncio.run(bar())
Traceback (most recent call last):
...
TypeError: 'async_generator' object is not an iterator
```

Do you want to open a separate issue?
